aws-sign-web.js

```diff
@@ -248,6 +248,42 @@
         }
     }
 
+    /**
+     * URI encode every byte. UriEncode() must enforce the following rules:
+     * URI encode every byte except the unreserved characters: 'A'-'Z', 'a'-'z', '0'-'9', '-', '.', '_', and '~'.
+     * The space character is a reserved character and must be encoded as "%20" (and not as "+").
+     * Each URI encoded byte is formed by a '%' and the two-digit hexadecimal value of the byte.
+     * Letters in the hexadecimal value must be uppercase, for example "%1A".
+     * Encode the forward slash character, '/', everywhere except in the object key name. For example, if the object key name is photos/Jan/sample.jpg, the forward slash in the key name is not encoded.
+     * Reference: http://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+     */
+    function uriEncode(input, encodeSlash) {
+        var result = [];
+        for (var i = 0; i < input.length; i++) {
+            var ch = input.charAt(i);
+            if ((ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z') || (ch >= '0' && ch <= '9') || ch === '_' || ch === '-' || ch === '~' || ch === '.') {
+                result.push(ch);
+            } else if (ch === '/') {
+                result.push(encodeSlash ? "%2F" : ch);
+            } else {
+                result.push(toHexUTF8(ch));
+            }
+        }
+        return result.join("");
+    }
+
```

Comment:
This doesn't work in (all) browsers and also is not the way to include dependencies. If we really want to introduce a new dependency then it should be added/required the same as `cryptoJs` above.
The focus of this library is to use plain-javascript (without any fancy transpiller or stuff...).

Also I'm not convinced if we really need a new dependency just for encoding stuff to UTF8.

Reply:
With your suggested enhancement to encodeURIComponent the new dependency is not needed. If you incorporate your (better) fix, I will close this Pull Request.